naga/src/proc/constant_evaluator.rs

```diff
@@ -1,9 +1,226 @@
+use std::iter;
+
+use arrayvec::ArrayVec;
+
 use crate::{
     arena::{Arena, Handle, UniqueArena},
     ArraySize, BinaryOperator, Constant, Expression, Literal, ScalarKind, Span, Type, TypeInner,
     UnaryOperator,
 };
 
+/// A macro that allows dollar signs (`$`) to be emitted by other macros. Useful for generating
+/// `macro_rules!` items that, in turn, emit their own `macro_rules!` items.
+///
+/// Technique stolen directly from
+/// <https://github.com/rust-lang/rust/issues/35853#issuecomment-415993963>.
+macro_rules! with_dollar_sign {
+    ($($body:tt)*) => {
+        macro_rules! __with_dollar_sign { $($body)* }
+        __with_dollar_sign!($);
+    }
+}
+
+macro_rules! gen_component_wise_extractor {
```

Comment:
I don't want to be negative, but the truth is that I cannot understand this at all.

Is there a way we could implement this without macro-generating macros?

Reply:
Using `cargo-expand`, most likely! :^)

I don't have any objection to inlining the output of the current macro. I'm cautious of DRY in the long term, but I don't think the cost will be noticeable until and unless we discover that the current body of this macro is incomplete or wrong. I actually consider this likely, but if you're willing to fix that sort of thing up, I don't see a reason to let perfect be the enemy of good (viz., merging this).

Reply:
Spoke with @jimblandy at length on a call here. Tentative conclusion: We'd like to iterate on this, but for now, we can merge this.